Fixing `set_start_value` only accept numbers (#2237)

henriquebecker91 opened the pull request:
Intends to close #2237 and some related open issues (#649, #794, #853).

Some guidance is needed. For now, I just edited what I found to be the relevant piece of documentation, and added a new tests as well flags to an old test (allowing more granular selection of which attributes are tested). I would like to know:

1. There are other places where the documentation needs to be changed? `JuMP` probably will need a change, but there are other points inside MOI?
2. Seems to me that allowing `nothing` to be passed is something done at the wrapper level, and the generic methods in MOI already allow it. The test will probably force the wrappers to allow this behavior.
3. How is the best way to check if the test is working? Should I download a bunch of wrappers and make they use my version of MOI, and then call their tests?

No hurry, @odow, I do not know if I will work during the week, but I will try finish it next weekend if I am sure of what needs to be done yet.

mlubin commented on src/attributes.jl:
```diff
@@ -965,6 +965,8 @@ struct VariableName <: AbstractVariableAttribute end
 
 A variable attribute for the initial assignment to some primal variable's value that the optimizer may use to warm-start the solve.
+
+The value `nothing` means the solver is free to decide which value to use. This attribute can be set back to `nothing` after being set to a numerical value.
```
The `VariablePrimalStart` is a hint that the optimizer "may" use to warm-start the solve when it's set. It's redundant to say the solver "is free to decide which value to use" when it's set to `nothing`, because this is also true when it's set to a number. Just say "May be a number or `nothing` (unset)."

henriquebecker91 replied:
Done.

odow commented:
1. We should update the documentation for `ConstraintDualStart` and `ConstraintPrimaStart` as well:
https://github.com/jump-dev/MathOptInterface.jl/blob/a6274e63f8cc020e16cb99a0b2d52bd8062e1613/src/attributes.jl#L1046-L1053

2. Rather than modifying the current test, you should add new tests
```Julia
function test_variable_primal_start(model, config)
    MOI.empty!(model)
    x = MOI.add_variable(model)
    @test MOI.get(model, MOI.VariablePrimalStart(), x) === nothing
    MOI.set(model, MOI.VariablePrimalStart(), x, 1.0)
    @test MOI.get(model, MOI.VariablePrimalStart(), x) == 1.0
    MOI.set(model, MOI.VariablePrimalStart(), x, nothing)
    @test MOI.get(model, MOI.VariablePrimalStart(), x) === nothing
end
```

3. You can add a test for the test here: https://github.com/jump-dev/MathOptInterface.jl/blob/master/test/Test/modellike.jl
Something like
```julia
model = MOIU.MockOptimizer(MOIU.UniversalFallback(MOIU.Model{Float64}()))
MOI.Test.test_variable_primal_start(model, MOI.Test.TestConfig())
```

henriquebecker91 commented:
@odow, you asked to do the same modification of documentation to `ConstraintPrimalStart` and `ConstraintDualStart`, ok, I did (not pushed yet), but then you suggest creating just a `function test_variable_primal_start(model, config)`, tests should be added for `ConstraintPrimalStart` and `ConstraintDualStart`, no?

henriquebecker91 commented:
I made a new commit. It updates the documentation of `VariablePrimalStart`, `ConstraintDualStart`, and `ConstraintPrimalStart` to what was suggested by @mlubin.

I also reverted the test I modified and instead created a new test, but not only for `VariablePrimalStart` (also for the constraint start values), and it is a little more involved than the test suggested by @odow.

I call the new test in the fashion suggested by @odow. The `config` parameter is not used, should it be used? 

My new test method has 5 commented `@test` calls. They fail if uncommented. I am not sure if this is desired behavior or not. I expect a decision from the maintainers about removing or uncommenting them. In the current situation (with these 5 commented) all tests pass.

henriquebecker91 commented:
Should I be doing something here?

henriquebecker91 commented:
Right now, I have a window of opportunity to make any changes necessary for this to be approved, while I wait for some reviews of my last paper. Any directions?

blegat commented on src/attributes.jl:
```diff
@@ -964,7 +964,7 @@ struct VariableName <: AbstractVariableAttribute end
     VariablePrimalStart()
 
-A variable attribute for the initial assignment to some primal variable's value that the optimizer may use to warm-start the solve.
+A variable attribute for the initial assignment to some primal variable's value that the optimizer may use to warm-start the solve. May be a number or nothing (unset).
```
nothing -> `nothing`

henriquebecker91 replied:
Already fixed in my local copy. Will be together with the fix of the other problems.

blegat commented on src/Test/modellike.jl:
```diff
@@ -386,6 +386,87 @@ function failcopytestca(dest::MOI.ModelLike)
+        MOI.set(model, vpattr, y, nothing)
+        MOI.set(model, cpattr, a, nothing)
+        MOI.set(model, cdattr, a, nothing)
+        MOI.set(model, cpattr, c, [nothing])
```
I would expect it to be `nothing`, not `[nothing]`

henriquebecker91 replied:
I will check this. I noticed now that I do not query the `cpattr` of `c` before setting it the first time. If the initial value of `cpattr` for `c` is `[nothing]`, then it would make sense keep it that way, to avoid breaking the reasoning that initially motivated this behavior, that is, ``if variables with an attribute unset return `nothing` when queried, then setting them with the returned value (i.e., nothing) should restore the original/unset state''. 

I have to admit I did not deal directly with this kinda of constraint, so I do not know what to expect here.
```julia
c = MOI.add_constraint(model, MOIU.operate(vcat, Float64, 2.0fz + 1.0), MOI.Nonnegatives(1))
```

henriquebecker91 replied:
Oh, okay, I was already checking it before and it has value `nothing`. So the `MOI.set` should take `nothing` and `MOI.get` should return `nothing`. I changed the tests and it works. Should we have some tests for `nothing` values inside the arrays that may be used to set this attribute, or setting the whole constraint to `nothing` and getting `nothing` back is enough?

henriquebecker91 replied:
Just changed the tests to expect `nothing` and set `nothing`. No test for `nothing` elements inside a Vector.

blegat replied:
It's ok if `nothing` element inside vector for a vector constraint is undefined behavior for now, this is out of the scope of this PR.

blegat commented on src/Test/modellike.jl:
```diff
@@ -386,6 +386,87 @@ function failcopytestca(dest::MOI.ModelLike)
+        MOI.set(model, cpattr, c, [nothing])
+        MOI.set(model, cdattr, c, [nothing])
+
+        #@test cpattr ∉ MOI.get(model, MOI.ListOfConstraintAttributesSet{F2, S2}())
```
Why is this commented ?

henriquebecker91 replied:
As I said above:

> My new test method has 5 commented @test calls. They fail if uncommented. I am not sure if this is desired behavior or not. I expect a decision from the maintainers about removing or uncommenting them. In the current situation (with these 5 commented) all tests pass.

henriquebecker91 replied:
Expanding a little on the above:

The documentation for [`ListOfVariableAttributesSet`](https://jump.dev/MathOptInterface.jl/stable/apireference/#MathOptInterface.ListOfVariableAttributesSet) and for [`ListOfConstraintAttributesSet`](https://jump.dev/MathOptInterface.jl/stable/apireference/#MathOptInterface.ListOfConstraintAttributesSet) are both not very clear (at least to me) on what happens if the attributes are unset.

* If I search for the attributes in these lists from a newly created model, they do not exist in the lists, as it is expected.
* If I search for them in these lists after setting them  in the model, they do exist in the lists, as it is expected.
* The problem is that after *unsetting* the attributes, they keep existing in these lists, I do not know if this is desired behaviour.

More specifically, I do not know if this is a list for every attribute that was ever set (i.e., `MOI.set` was called with it), or just for the attributes with a value different from `nothing` at the moment the list is queried.

blegat replied:
Yes, this is a bit ambiguous, we don't want the definition to be to constrained here as it would make it difficult to implement and it does not change anything for the main use case which is using it for copying.
You can remove these tests or add a comment explaining that it currently fails but we may want to enable these in the future.

henriquebecker91 replied:
I kept them commented and added a comment explaining why they are commented referencing this thread.

henriquebecker91 commented:
I followed all @blegat suggestions and did a minor cosmetic change (changed `!(... \in ...)` to `... \notin ...`).

henriquebecker91 commented:
Screenshot of tests passing:
[![2020-09-29-11-52-03-1920x1080.png](https://i.postimg.cc/s2GdZBQV/2020-09-29-11-52-03-1920x1080.png)](https://postimg.cc/cKW5pLBj)

blegat commented on src/Test/modellike.jl:
```diff
@@ -386,6 +386,100 @@ function failcopytestca(dest::MOI.ModelLike)
+        # determined behaviour, there are two possibilities about the behaviour
+        # of ListOfConstraintAttributesSet and ListOfVariableAttributesSet:
+        #
+        # 1) They list all attributes ever set, even after unsetted.
```
unsetted -> unset

blegat reviewed (approved):
Can you remove the "Draft" state of the PR ?

henriquebecker91 commented:
Sorry, forgot about that. Now it is ready for review. I will make that small comment change.
